snow/block.go

```diff
@@ -0,0 +1,348 @@
+			zap.Stringer("blkID", b.Input.ID()),
+		)
+	default:
+		b.vm.log.Info("Verifying block", zap.Stringer("block", b))
+		if err := b.innerVerify(ctx); err != nil {
+			return err
+		}
+
+		if err := event.NotifyAll[O](ctx, b.Output, b.vm.verifiedSubs...); err != nil {
+			return err
+		}
+	}
+
+	b.vm.verifiedL.Lock()
+	b.vm.verifiedBlocks[b.Input.ID()] = b
+	b.vm.verifiedL.Unlock()
+
+	if b.verified {
+		b.vm.log.Debug("verified block",
+			zap.Stringer("blk", b.Output),
+			zap.Bool("ready", ready),
+		)
+	} else {
+		b.vm.log.Debug("skipped block verification",
+			zap.Stringer("blk", b.Input),
+			zap.Bool("ready", ready),
+		)
+	}
+	return nil
+}
+
+func (b *StatefulBlock[I, O, A]) innerVerify(ctx context.Context) error {
+	// Fetch my parent to verify against
+	parent, err := b.vm.GetBlock(ctx, b.Parent())
+	if err != nil {
+		return err
+	}
+
+	// If my parent has not been verified and we're no longer in dynamic state sync,
+	// we must be transitioning to normal consensus.
+	// Attempt to verify from the last accepted block through to this block to
+	// compute my parent's Output state.
+	if !parent.verified {
+		b.vm.log.Info("parent was not verified during innerVerify",
+			zap.Stringer("lastAccepted", b.vm.lastAcceptedBlock),
+			zap.Stringer("parent", parent),
+			zap.Stringer("block", b),
+		)
+		blksToProcess, err := b.vm.getExclusiveBlockRange(ctx, b.vm.lastAcceptedBlock, b)
+		if err != nil {
+			return err
+		}
+		b.vm.log.Info("found blks to process",
+			zap.Int("numBlks", len(blksToProcess)),
+			zap.Stringer("first", blksToProcess[0]),
+			zap.Stringer("last", blksToProcess[len(blksToProcess)-1]),
+		)
+		parent := b.vm.lastAcceptedBlock
+		for _, ancestor := range blksToProcess {
+			if err := ancestor.verify(ctx, parent.Output); err != nil {
+				return err
+			}
+			parent = ancestor
+		}
+		b.vm.log.Info("finished verified ancestors")
+	} else {
+		b.vm.log.Info("parent was already verified")
+	}
+
+	// Verify the block against the parent
+	return b.verify(ctx, parent.Output)
+}
+
+// markAccepted marks the block and updates the required VM state.
+// iff parent is non-nil, it will request the chain to Accept the block.
+// The caller is responsible to provide the accepted parent if the VM is in a ready state.
+func (b *StatefulBlock[I, O, A]) markAccepted(ctx context.Context, parent *StatefulBlock[I, O, A]) error {
+	if err := b.vm.inputChainIndex.UpdateLastAccepted(ctx, b.Input); err != nil {
+		return err
+	}
+
+	if parent != nil {
+		if err := b.accept(ctx, parent.Accepted); err != nil {
+			return err
+		}
+	}
+
+	b.vm.verifiedL.Lock()
+	delete(b.vm.verifiedBlocks, b.Input.ID())
+	b.vm.verifiedL.Unlock()
+
+	b.vm.setLastAccepted(b)
+
+	return b.notifyAccepted(ctx)
+}
+
+func (b *StatefulBlock[I, O, A]) notifyAccepted(ctx context.Context) error {
+	// If I was not actually marked accepted, notify pre ready subs
+	if !b.accepted {
+		return event.NotifyAll(ctx, b.Input, b.vm.preReadyAcceptedSubs...)
+	}
```

Comment:
Instead of having two separate subscriptions for ready/preready, we could just wrap the event type and add the `accepted` field so that the downstream consumer can decide how to handle the event based on the accepted field

Reply:
The subscriptions receive different types because prior to completing dynamic state sync the blocks are not actually verified/accepted, so it can only send the input type.

In practice, a pre-ready subscription plays a different role than an accepted subscription, so I think separately registered subscriptions makes sense here.